photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,307 @@
+
+/**
+ * A DataReader implementation that reads the Avro format
+ */
+class AvroDataReader(sc: SparkContext) extends DataReader {
+
+  import AvroDataReader._
+
+  /**
+   * Internal sql context
+   */
+  private val sqlContext = new SQLContext(sc)
+
+  /**
+   * Reads the avro file at the given path into a DataFrame, generating a default index map for feature names
+   *
+   * @param path the path to the avro file or folder
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the loaded and transformed DataFrame
+   */
+  override def read(
+      path: String,
+      featureColumnMap: Map[String, Set[String]]): (DataFrame, Map[String, IndexMapLoader]) = {
+
+    val data = sqlContext.read.avro(path)
+    val merged = mergeColumns(data, featureColumnMap)
+    val indexMapLoaders = generateIndexMapLoaders(merged, featureColumnMap.keys.toSet)
+
+    val transformed = vectorize(
+      merged,
+      indexMapLoaders)
+
+    (transformed, indexMapLoaders)
+  }
+
+  /**
+   * Reads the avro file at the given path into a DataFrame, using the given index map for feature names
+   *
+   * @param path the path to the avro file or folder
+   * @param indexMapLoaders a map of index map loaders, containing one loader for each feature column
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the loaded and transformed DataFrame
+   */
+  override def read(
+      path: String,
+      indexMapLoaders: Map[String, IndexMapLoader],
+      featureColumnMap: Map[String, Set[String]]): DataFrame = {
+
+    val data = sqlContext.read.avro(path)
+
+    vectorize(
+      mergeColumns(data, featureColumnMap),
+      indexMapLoaders)
+  }
+
+  /**
+   * Generates a temporary column name
+   *
+   * @param name the base name for the column
+   * @return the temporary column name
+   */
+  protected def tempColumnName(name: String): String = s"__${name}__${System.currentTimeMillis}__"
+
+  /**
+   * Read a feature tuple from the name, term, and value DataFrame row
+   *
+   * @param row the row from which to read the name, term, value tuple
+   * @return the feature tuple
+   */
+  protected def readFeature = (feature: Row) => {
```

Comment:
I've done it this way so I can include the function in the closure on line 133 rather than serializing the entire object. As I write this, though, I realize that I could accomplish the same thing by making this function a member of the companion object. I think I'll do that instead.